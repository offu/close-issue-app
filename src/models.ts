/** Config for issue matching. */
export interface IssueConfig {
  /** content to match */
  content: Array<string>
}

/** Config for the issue bot */
export interface BotConfig {
  /** One BotConfig can include several IssueConfig for different issues */
  issueConfigs: Array<IssueConfig>
  /** Comment that will be sent if an issue is judged to be closed */
  comment: string,
  /** whether the keywords are case-sensitive */
  caseSensitive?: boolean
}

/**
 * check something if it is an IssueConfig
 */
export function isIssueConfig (item: any): item is IssueConfig {
  if (!('content' in item
  && Array.isArray(item.content))) {
    return false
  }
  return item.content.every((i) => {
    return typeof i === 'string'
  })
}
/**
 * check something if it is a BotConfig
 */
export function isBotConfig (item: any): item is BotConfig {
  if (!(item !== null
  && typeof item === 'object'
  && 'comment' in item
  && 'issueConfigs' in item
  && typeof item.comment === 'string'
  && Array.isArray(item.issueConfigs))) {
    return false
  }
  return item.issueConfigs.every(isIssueConfig)
}
/**
 * inject default values into config
 */
export function botConfigFactory (config: BotConfig): BotConfig {
  if (!('caseSensitive' in config)) {
    config.caseSensitive = true
  }
  return config
}

export const errorComment: string = 'The app gets an error. :('
