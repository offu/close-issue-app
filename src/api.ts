import { Context } from 'probot'
import { Buffer } from 'buffer'

export async function closeIssue (context: Context) {
  const params = {
    owner: context.payload.repository.owner.login,
    repo: context.payload.repository.name,
    issue_number: context.payload.issue.number
  }
  await context.github.issues.update({
    ...params,
    state: 'closed'
  })
}

export async function createLabelIfNotExist (context, label: string) {
  try {
    const l = await context.github.issues.getLabel({
      owner: context.payload.repository.owner.login,
      repo: context.payload.repository.name,
      name: label
    })
  } catch (e) {
    if (e.hasOwnProperty('status') && e.status === 404) {
      await context.github.issues.createLabel({
        owner: context.payload.repository.owner.login,
        repo: context.payload.repository.name,
        name: label,
        color: 'aa1414'
      })
    } else {
      throw e
    }
  }
}

export async function addLabel (context: Context, label: string) {
  await createLabelIfNotExist(context, label)
  await context.github.issues.addLabels({
    owner: context.payload.repository.owner.login,
    repo: context.payload.repository.name,
    number: context.payload.issue.number,
    labels: [label]
  })
}

export async function createComment (context: Context, comment: string) {
  const params = {
    owner: context.payload.repository.owner.login,
    repo: context.payload.repository.name,
    number: context.payload.issue.number
  }
  await context.github.issues.createComment({
    body: comment,
    ...params
  })
}

export async function getContent (context: Context, path: string) {
  const params = {
    owner: context.payload.repository.owner.login,
    repo: context.payload.repository.name
  }
  const resp = await context.github.repos.getContents({
    path,
    ...params
  })
  return Buffer.from(resp.data[0].content, resp.data[0].encoding).toString('utf-8')
}
