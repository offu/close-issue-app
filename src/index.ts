interface ProBot {
  log: any
}

export = (robot: ProBot) => {
  robot.log('Yay, the app was loaded!')
}
