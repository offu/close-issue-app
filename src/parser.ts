import * as yaml from 'js-yaml'
import { BotConfig, isBotConfig } from './models'

/**
 * Parse the file according to the path into BotConfig
 */
export function parseConfig (content: string): BotConfig {
  const data = yaml.safeLoad(content)
  if (!isBotConfig(data)) {
    throw new Error('invalid config')
  } else {
    return data
  }
}

export function shouldClose (botConfig: BotConfig, content: string): boolean {
  return !botConfig.issueConfigs.some((issueConfig) => {
    return issueConfig.content.every((issueContent) => {
      if (botConfig.caseInsensitive) {
        return content.toLowerCase().includes(issueContent.toLowerCase())
      } else {
        return content.includes(issueContent)
      }
    })
  })
}
