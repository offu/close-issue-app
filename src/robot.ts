import { Robot } from 'probot'

export = (robot: Robot) => {
  robot.log('Yay, the app was loaded!')
}
