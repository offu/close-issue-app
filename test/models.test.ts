import { isBotConfig, isIssueConfig } from '../src/models'
import * as path from 'path'
import exampleConfig from './example.config.json'

describe('isBotConfig', () => {
  it('example config', () => {
    expect(isBotConfig(exampleConfig)).toBeTruthy()
  })

  it('empty example', () => {
    const emptyExample = {}
    expect(isBotConfig(emptyExample)).toBeFalsy()
  })

  it('empty issues example', () => {
    const wrongExample1 = { issueConfigs: [], comment: 'test' }
    expect(isBotConfig(wrongExample1)).toBeTruthy()
  })

  it('no issues example', () => {
    const wrongExample2 = { comment: 'test' }
    expect(isBotConfig(wrongExample2)).toBeFalsy()
  })

  it('right example', () => {
    const rightExample = {
      issueConfigs: [{ content: ['123'] }],
      comment: 'test'
    }
    expect(isBotConfig(rightExample)).toBeTruthy()
  })
})

describe('isIssueConfig', () => {
  it('example issue config', () => {
    expect(isIssueConfig(exampleConfig.issueConfigs[0])).toBeTruthy()
  })

  it('no content example', () => {
    const emptyExample = {}
    expect(isIssueConfig(exampleConfig)).toBeFalsy()
  })

  it('empty content example', () => {
    const emptyItemsExample = { content: [] }
    expect(isIssueConfig(emptyItemsExample)).toBeTruthy()
  })

  it('wrong content example', () => {
    const wrongItemsExample = { content: [123] }
    expect(isIssueConfig(wrongItemsExample)).toBeFalsy()
  })
})
