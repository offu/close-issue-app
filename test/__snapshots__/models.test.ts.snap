// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`botConfigFactory config with caseSensitive 1`] = `
Object {
  "caseSensitive": false,
  "comment": "test",
  "issueConfigs": Array [
    Object {
      "content": Array [
        "123",
      ],
    },
  ],
}
`;

exports[`botConfigFactory config without caseSensitive 1`] = `
Object {
  "caseSensitive": true,
  "comment": "test",
  "issueConfigs": Array [
    Object {
      "content": Array [
        "123",
      ],
    },
  ],
}
`;
